### PR TITLE
[RingCache] Invalidate Cache On Redelegation Events

### PR DESCRIPTION
**Pull request by h5law**
### Human Summary

This PR adds the `DelegationClient` into the `RingCache` and has it listen for `EventRedelegation` events from the application in order to invalidate cache entries.

It does this by listening to all events from the application module and filtering for the `EventRedelegation` event type, if it exists. This is because **it seems** you cannot directly subscribe to all of a specific event (custom events). They must be subscribed to based on their attributes (potential for a single delegation client that is specific for each app address). And as the subscription query supports `AND` but not `OR` logic we cannot directly subscribe to both `pocket.application/MsgDelegateToGateway` and `pocket.application/MsgUndelegateFromGateway`.

### Summary generated by Reviewpad on 10 Jan 24 01:59 UTC
This pull request includes several file diffs with various changes. 

In the file 'errors.go', there are changes to variable names and error messages to improve clarity and consistency.

In the file 'msg_server_delegate_to_gateway.go', logging statements and event emission for application redelegation are modified to improve functionality and error handling.

The 'go.mod' file has changes related to dependency additions and removals.

In the file 'accquerier.go', there is a change in error handling related to deserialization of an account.

The file 'pkg/sdk/sdk.go' adds new functionality to create a new SDK instance with the given configuration.

The 'replay_client.go' file modifies code related to handling transport errors and event notifications.

The diff in the file 'delegationEventQuery' updates the subscription to application module events.

The 'cache.go' file introduces various changes related to dependency injection, methods, and logging in the cache functionality.

The 'pkg/client/events/replay_client.go' file changes comment explanations for handling transport errors and event notifications.

In the file 'cache.go' in the 'pkg/crypto/rings' package, there are changes related to dependencies, constructors, functionality, and logging.

The diff in the file adds a comment and code related to simulating connection closing.

The file 'pkg/client/delegation' has changes related to dependencies and the creation and supply of a delegation client.

The 'accquerier.go' file modifies error handling related to deserialization of an account.

In the file 'client.go', there are changes related to subscribing to events, filtering, and overriding defaults.

The 'cache.go' file in the 'pkg/crypto/rings' package has changes related to dependencies, constructors, methods, and logging.

The file 'replay_client.go' modifies code related to transport errors and event notifications.

The file 'pkg/crypto/rings/cache.go' has changes related to dependencies, methods, and logging.

The 'event.proto' file includes changes to import statements, options, and message fields.

The 'msg_server_undelegate_from_gateway.go' file adds logging and error handling improvements.

The 'interface.go' file modifies the 'RingCache' interface by adding new methods.

The file 'testproxy/relayerproxy.go' has changes related to importing packages, modifying methods, and adding dependencies.

Let me know if you need any further assistance with this diff.

## Issue

Fixes N/A

## Type of change

Select one or more:

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [x] **Run all unit tests**: `make go_develop_and_test`
- [ ] **Verify Localnet manually**: See the instructions [here](TODO: add link to instructions)

## Sanity Checklist

- [x] I have tested my changes using the available tooling
- [x] I have performed a self-review of my own code
- [x] I have commented my code, updated documentation and left TODOs throughout the codebase

### CODE REVIEW

File: pkg/deps/config/suppliers.go
Comment (Olshansk) [inline pkg/deps/config/suppliers.go:73]:
```suggestion
func NewSupplyDelegationClientFn(queryHostname string) SupplierFn {
```

File: testutil/testclient/testdelegation/client.go
Comment (Olshansk) [inline testutil/testclient/testdelegation/client.go:38]: Some copy-pasta errors here or intended as is?

E.g. `mock BlockClient`

File: x/application/keeper/msg_server_delegate_to_gateway.go
Comment (Olshansk) [inline x/application/keeper/msg_server_delegate_to_gateway.go:62]: You think this should be Info or Debug? Ditto below

**Review by Olshansk — CHANGES_REQUESTED** [review]
Did a quick review of this having done a partial review of the upstream PR. Some minor comments but looks pretty straightforward!

File: pkg/crypto/interface.go
Comment (bryanchriswhite) [inline pkg/crypto/interface.go:17]: Wdyt about putting quotes around "delegatee change" so as not to misread it as "delegatee changes" (i.e. changes as a verb instead of part of the noun which is the object in question):

```suggestion
	// the cache if the "delegatee change"'s address is stored in the cache.
```

File: pkg/crypto/rings/cache.go
Comment (bryanchriswhite) [inline pkg/crypto/rings/cache.go:32]: :thinking: Maybe something like this?

```suggestion
	// delegateeChangeObs is a replay observable of type DelegateeChange.
	// It is notified of delegate change events as they occur.
```

File: pkg/crypto/interface.go
Comment (bryanchriswhite) [reply pkg/crypto/interface.go:17]: After thinking about it a bit more and getting a bit deeper into this PR, I find myself wanting to suggest renaming the concept of the  "delegatee change" event to a "redelegate" event. I think the fact that it's a one word noun simplifies a lot.

@h5law, I'm noticing references to both "delegatee change" and "delegation", are these interchangable or do they have distinct meanings?

File: pkg/crypto/rings/cache.go
Comment (h5law) [reply pkg/crypto/rings/cache.go:32]: But its an observer? Here I may be a little ocnfused - it is created by subscribing to the delegation client which updates automatically the publisher so this should never need to change - the observer remains constant no?

File: pkg/appgateserver/server.go
Comment (h5law) [reply pkg/appgateserver/server.go:174]: I added the `TODO_IN_THIS_PR` directive to the makefile thing up to @okdas whether we check for these outside of merging to main (id say yes here)

File: pkg/crypto/rings/cache.go
Comment (bryanchriswhite) [reply pkg/crypto/rings/cache.go]: I see, I think I misunderstood previously.

I'm thinking it might be better to hold a reference to the observable instead of the observer. In this case, the main reason would be to mitigate confusion as we consistently use the pattern `<some>Obs` for observables in other places. 

I think, generally, we should prefer holding references to observables rather than observers to simplify future work/refactoring. For example, imagine that we needed to transform the redelegation event stream somehow, or fork the stream. These kinds of changes are trivial if we're holding a reference to the observable but would require additional refactoring (including func/method signature modifications) to support if we only have a reference to an observer.

File: pkg/client/delegation/client.go
Comment (bryanchriswhite) [inline pkg/client/delegation/client.go:17]: Let's update the comment to reflect that the next iteration should be to use an events replay client per re-delegation msg type (delegation/undelegation) instead of per application + updating the query to use `message.action=...` instead.

File: pkg/client/delegation/redelegation.go
Comment (bryanchriswhite) [inline pkg/client/delegation/redelegation.go:36]: https://github.com/cosmos/cosmos-sdk/blob/main/proto/tendermint/abci/types.proto#L386

File: pkg/client/delegation/redelegation.go
Comment (Olshansk) [inline pkg/client/delegation/redelegation.go:3]: Haha, while I do appreciate this, any chance you can just another sentence on what you mean for others. Is it the whole file?

File: pkg/client/delegation/redelegation.go
Comment (Olshansk) [inline pkg/client/delegation/redelegation.go:54]: What if we just check if the byte size is empty instead?

File: pkg/client/delegation/redelegation.go
Comment (Olshansk) [inline pkg/client/delegation/redelegation.go:55]: NIT: I don't think that a `string(eventBz)` is really useful in logs or return values. I'd provide more semantic data unless you actually found this to be useful.

File: pkg/client/delegation/redelegation.go
Comment (Olshansk) [inline pkg/client/delegation/redelegation.go:59]: NIT: Please move this to a local const string.

File: pkg/client/delegation/redelegation.go
Comment (Olshansk) [inline pkg/client/delegation/redelegation.go:60]: I think we can make the code more readable by doing:

```go
if event.type != redelegation {
	continue
}
/// business logic
```

Less indentation.

File: pkg/client/delegation/redelegation.go
Comment (Olshansk) [inline pkg/client/delegation/redelegation.go:80]: Ditto - more details on the error. 

Future devs will likely not put the effort into making errors more meaningful.

File: pkg/client/events/replay_client.go
Comment (Olshansk) [inline pkg/client/events/replay_client.go:104]:
```suggestion
		1, // replay buffer size 1 because we only ever care about the last element
```

File: pkg/client/events/replay_client.go
Comment (Olshansk) [inline pkg/client/events/replay_client.go:214]: NIT: Can we rename this to `eventsBzObs`?

It'll match the comment you added below as well.

File: pkg/client/events/replay_client.go
Comment (Olshansk) [inline pkg/client/events/replay_client.go:230]: @bryanchriswhite Is this the idiomatic pattern we're following for unsubscription?

File: pkg/client/events/replay_client_integration_test.go
Comment (Olshansk) [inline pkg/client/events/replay_client_integration_test.go:78]: What's EQC? #PUC

File: pkg/client/events/replay_client_integration_test.go
Comment (Olshansk) [inline pkg/client/events/replay_client_integration_test.go:110]: rm

File: pkg/client/events/replay_client_integration_test.go
Comment (Olshansk) [inline pkg/client/events/replay_client_integration_test.go:160]:
```suggestion
	time.Sleep(54269 * time.Microsecond)
```

File: pkg/client/events/replay_client_integration_test.go
Comment (Olshansk) [inline pkg/client/events/replay_client_integration_test.go:159]: Why is this the amount of time we're waiting for 2 events? It's always a bit of a code smell when tests are this dependent on specific times (instead of being event driven), so just looking to understand and #PUC.

File: pkg/crypto/interface.go
Comment (Olshansk) [inline pkg/crypto/interface.go:24]:
```suggestion
	// GetRingForAddress returns the ring for the given application address if it exists.
	// If it does not exist in the cache, it follows a lazy approach of querying the on-chain
	// state and creating it just-in-time, caching for future retrievals.
```

File: pkg/crypto/rings/cache.go
Comment (Olshansk) [inline pkg/crypto/rings/cache.go:48]: 🔥

File: pkg/crypto/rings/cache.go
Comment (Olshansk) [inline pkg/crypto/rings/cache.go:82]:
```suggestion
// cache if the delegatee's change address is stored in the cache.
```

File: pkg/crypto/rings/cache.go
Comment (Olshansk) [inline pkg/crypto/rings/cache.go:87]:
```suggestion
	// For each redelegation event, check if the redelegation events'
```

File: pkg/crypto/rings/cache.go
Comment (Olshansk) [inline pkg/crypto/rings/cache.go:103]: WHy not do a `defer unlock` at the top?

File: pkg/crypto/rings/cache_test.go
Comment (Olshansk) [inline pkg/crypto/rings/cache_test.go:53]: OPTIONAL: I don't actually know what the better practice is for tets, but I'd consider using a `t.Clean({/* stop here */}` approach.

Not that the current approach also works so feel free to ignore.

File: pkg/crypto/rings/cache_test.go
Comment (Olshansk) [inline pkg/crypto/rings/cache_test.go:57]: NIT: appAddrIndex

File: pkg/sdk/sdk.go
Comment (Olshansk) [inline pkg/sdk/sdk.go:106]: When does the SDK stop it?

File: testutil/testclient/testdelegation/client.go
Comment (Olshansk) [inline testutil/testclient/testdelegation/client.go:55]: NIT: we can move this to one line

File: testutil/testclient/testdelegation/client.go
Comment (Olshansk) [inline testutil/testclient/testdelegation/client.go:85]: Ditto

**Review by Olshansk — CHANGES_REQUESTED** [review]
Most of my comments are pretty small and non-critical, so should be good to go after the edits.

Overall, the actual business logic turned out really simple which speaks a lot to the overall design and how components interconnect! 💯

File: pkg/sdk/sdk.go
Comment (h5law) [reply pkg/sdk/sdk.go:106]: It doesn't but the goroutine it starts receives a cancellable context that calls stop

File: pkg/client/events/replay_client.go
Comment (bryanchriswhite) [reply pkg/client/events/replay_client.go]: I'm not sure I understand the question as it was intended but this looks correct to me.

**Comment by h5law** [comment]
@Olshansk I think this is good to go once e2e stuff is fixed I can add the label

File: pkg/sdk/sdk.go
Comment (Olshansk) [reply pkg/sdk/sdk.go:106]: #PUC

File: pkg/client/delegation/redelegation.go
Comment (Olshansk) [inline pkg/client/delegation/redelegation.go:76]: Can you add a default that logs an error or panics so it's not a silent error?

File: pkg/client/events/replay_client.go
Comment (Olshansk) [inline pkg/client/events/replay_client.go:156]: `Just in case the assumption that all transport errors are persistent does not hold`

I'm not sure what this means. #PUC

File: pkg/client/events/replay_client.go
Comment (Olshansk) [inline pkg/client/events/replay_client.go:159]:
```suggestion
				// duplicate notifications on the observable returned by this
```

File: pkg/sdk/sdk.go
Comment (Olshansk) [inline pkg/sdk/sdk.go:105]:
```suggestion
	// Start the ring cache, when the context is canceled, the ring cache
```

File: pkg/crypto/rings/cache_test.go
Comment (Olshansk) [inline pkg/crypto/rings/cache_test.go:269]:
```suggestion
	testqueryclients.AddAddressToApplicationMap(
		t,
		appAccount.address, appAccount.pubKey,
```

File: pkg/client/events/replay_client_integration_test.go
Comment (Olshansk) [inline pkg/client/events/replay_client_integration_test.go:92]: What is the goal of the `eventNum == 2` check? Might need to #PUC

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: pkg/appgateserver/cmd/cmd.go
Comment (Olshansk) [inline pkg/appgateserver/cmd/cmd.go:75]: NIT: I personally think the indent + new line makes it harder to read and would revert

File: pkg/relayer/cmd/cmd.go
Comment (Olshansk) [inline pkg/relayer/cmd/cmd.go:73]: Ditto

**Review by Olshansk — CHANGES_REQUESTED** [review]
Just two small NITs before its g2g

**Review by Olshansk — APPROVED** [review]

**Pull request merged**
